### PR TITLE
conf: change suite.rc to flow.cylc

### LINKED ISSUE #3689: conf: configuration file changes
As documented in the [config change proposal document](https://cylc.github.io/cylc-admin/proposal-config-changes.html#config-file-naming):

- [x] The `.cylc` extension and the new configuration file names.
  - [x] With support for the deprecated `suite.rc` filename with a warning of some kind for the user.
- [ ] The new global config hierarchy.
- [ ] Any changes to `CYLC_CONF_PATH` required.
- [ ] The new `CYLC_SITE_CONF_PATH` variable.

**Pull requests welcome!**

### PR DESCRIPTION
**Pull request by MetRonnie**
These changes partially address #3689 

Best to look at this one commit by commit. The commits called "Wider rename..." usually consist of the same change done across many files, so you probably only need to take a look at a couple of the files in those.

### Summary
- `suite.rc --> flow.cylc`
- `flow.rc & global.rc --> global.cylc`
- The directory `log/suiterc/ --> log/flow-config/`
- Generally, `suiterc` in variable names `--> flow_config`, and generally `rc` in variable names `--> config`
  - Functional tests: `test_header` function `create_test_globalrc --> create_test_global_config`
  - Integration tests: (in `tests/i/utils/flow_writer.py`): `flow_writer.suiterc() --> flow_writer.flow_config_str()`
  - When referring to a file or filename/path (rather than the config itself), generally `suiterc --> flow_file`
- Every other `*.rc --> *.cylc`

### Backwards compatibility
- The first time a user `cylc run`s a directory that contains only `suite.rc`, it will automatically create a symlink `flow.cylc` to that `suite.rc`.
- If a user `cylc run`s an already-registered suite/workflow and `flow.cylc` doesn't exist (i.e. suite/workflow was run on a previous version of Cylc), it will re-register 

(No backwards compatibility for flow.rc)

---

**Requirements check-list**
- [x] I have read `CONTRIBUTING.md` and added my name as a Code Contributor.
- [x] Contains logically grouped changes (else tidy your branch by rebase).
- [x] Does not contain off-topic changes (use other PRs for other changes).

- [x] Tests for "upgrading" existing `suite.rc` to `flow.cylc` included

- [x] Appropriate change log entry included.

- [x] (master branch) I have opened a documentation PR at https://github.com/cylc/cylc-doc/pull/142

- [x] No dependency changes.

### CODE REVIEW

File: cylc/flow/suite_files.py [hunk 1]
```diff
@@ -449,15 +453,26 @@ def register(reg=None, source=None, redirect=False, rundir=None):
...
+        suite_rc_path = os.path.join(source, SuiteFiles.SUITE_RC)
+        if os.path.isfile(suite_rc_path):
+            os.symlink(suite_rc_path, flow_file_path)
+            LOG.warning(
+                f'The filename "{SuiteFiles.SUITE_RC}" is deprecated in favor '
+                f'of "{SuiteFiles.FLOW_FILE}". The latter has been '
+                f'automatically created as a symlink to the former.')
+        else:
+            raise SuiteServiceFileError(
+                f"no flow.cylc or suite.rc in {source}")
```
Comment (MetRonnie) [inline cylc/flow/suite_files.py:475]: What do people think of this method of "upgrading" existing `suite.rc` to `flow.cylc`?

Comment (oliver-sanders) [reply cylc/flow/suite_files.py:475]: Seems legit, the symlink isn't strictly necessary but it's not a problem or overhead and saves us replicating this logic elsewhere (e.g. `cylc scan`) so fine by me.

Comment (hjoliver) [reply cylc/flow/suite_files.py:475]: Seems fine to me too :+1:

File: cylc/flow/suite_files.py [hunk 2]
```diff
@@ -449,15 +453,26 @@ def register(reg=None, source=None, redirect=False, rundir=None):
...
+    flow_file_path = os.path.join(source, SuiteFiles.FLOW_FILE)
+    if not os.path.isfile(flow_file_path):
+        # If using deprecated suite.rc, symlink it into flow.cylc:
+        suite_rc_path = os.path.join(source, SuiteFiles.SUITE_RC)
+        if os.path.isfile(suite_rc_path):
+            os.symlink(suite_rc_path, flow_file_path)
+            LOG.warning(
+                f'The filename "{SuiteFiles.SUITE_RC}" is deprecated in favor '
+                f'of "{SuiteFiles.FLOW_FILE}". The latter has been '
+                f'automatically created as a symlink to the former.')
```
Comment (hjoliver) [inline cylc/flow/suite_files.py:472]: Warning message overly verbose?
```suggestion
                f'of "{SuiteFiles.FLOW_FILE}". Symlink created.')
```

Comment (oliver-sanders) [reply cylc/flow/suite_files.py]: Might be an idea to give a deprecation timeframe in the warning, something like:

```
Warning: The suite.rc filename as been deprecated by flow.cylc.
support for suite.rc will be withdrawn in Cylc9.
```

(presumably we should be good to kill this support by Cylc9?)

File: cylc/flow/cfgspec/globalcfg.py [hunk 3]
```diff
@@ -87,21 +87,21 @@
```
[the remark this answers is not in the document]
Comment (MetRonnie) [reply cylc/flow/cfgspec/globalcfg.py:87]: Do you mean `flow.rc`? Would it make sense to split that off into another PR?

[the remark this answers is not in the document]
Comment (oliver-sanders) [reply cylc/flow/cfgspec/globalcfg.py:87]: Ah, just read the PR description, I would have thought it would be easier to do both at the same time, but meh.

**Comment by oliver-sanders** [comment]
> If a user cylc runs an already-registered suite/workflow and flow.cylc doesn't exist (i.e. suite/workflow was run on a previous version of Cylc), it will re-register

> is this overkill

No, it's a really valuable feature.

**Comment by MetRonnie** [comment]
After this is merged, would it make sense to add a GH Actions workflow/step that checks PRs' additions don't include `suite.rc` etc going forward, in case of old habits dying hard?

**Comment by hjoliver** [comment]
> The directory `log/suiterc/` --> `log/flow_config/`

A very minor point, but the `cylc-run` directory uses `-` rather than `_`, so for consistency I'd prefer `flow-config` here.

Actually maybe just `config` or `conf` would do - we're not going to be storing other config info here are we?

**Comment by MetRonnie** [comment]
> maybe just config or conf would do - we're not going to be storing other config info here are we?

As a relative newcomer, I feel that including "flow" in the name makes its purpose clearer.

**Comment by hjoliver** [comment]
> As a relative newcomer, I feel that including "flow" in the name makes its purpose clearer.

It is already under the `cylc-run` directory, which narrows down the possibilities quite a bit.  But I'm not averse to `flow-config` (with a hyphen as explained above) if that's the consensus.

**Comment by oliver-sanders** [comment]
I would probably go for either:

* `log/config/<timestamp>-(run|restart|reload).cylc`
* `log/config/<timestamp>-(run|restart|reload)-flow.cylc`

**Comment by MetRonnie** [comment]
I've changed it to `flow-config/`, I hope that's ok because it's too 🌡️ hot today to be trying to rebase anymore 😝

File: cylc/flow/scripts/cylc_graph.py [hunk 4]
```diff
@@ -27,7 +27,7 @@
 from cylc.flow.cycling.loader import get_point
 from cylc.flow.exceptions import UserInputError, SuiteServiceFileError
 from cylc.flow.option_parsers import CylcOptionParser as COP
-from cylc.flow.suite_files import get_suite_rc
+from cylc.flow.suite_files import get_flow_file
```
[the remark this answers is not in the document]
Comment (hjoliver) [reply cylc/flow/scripts/cylc_graph.py:30]: And that'll be a job :grimacing: :
```
$ git grep -i suite | wc -l
7775
```

[the remark this answers is not in the document]
Comment (oliver-sanders) [reply cylc/flow/scripts/cylc_graph.py:30]: Yeah, I wouldn't worry about renaming Python modules in this PR (there are others), we will have to do a right stint of renaming, but that can wait a little while.

File: cylc/flow/parsec/README.md [hunk 5]
```diff
@@ -28,7 +28,7 @@ defines the "file format" (next).
 
 ### File spec
 
-**E.g. ``cylc/flow/parsec/test_spec.py`` for ``cylc/flow/parsec/test.rc`` and Cylc file
+**E.g. ``cylc/flow/parsec/test_spec.py`` for ``cylc/flow/parsec/test.cylc`` and Cylc file
```
[the remark this answers is not in the document]
Comment (MetRonnie) [reply cylc/flow/parsec/README.md:31]: Well, I figured from now on there shouldn't be any mention of `.rc`

**Comment by MetRonnie** [comment]
Just pushed one last change because commands like `cylc validate .` weren't working for a nested suite like `~/cylc-run/foo/bar/` where only `suite.rc` is present

File: tests/unit/test_pathutil.py [hunk 6]
```diff
@@ -16,14 +16,12 @@
...
 
 from unittest import TestCase
-from unittest.mock import call, patch, MagicMock
+from unittest.mock import patch, MagicMock
 
 import pytest
 import os
 import logging
 
-from tests.unit.conftest import mock_glbl_cfg
```
Comment (MetRonnie) [inline tests/unit/test_pathutil.py:25]: No need to import from `conftest.py`, pytest does this automatically it seems

Comment (oliver-sanders) [reply tests/unit/test_pathutil.py:25]: It does indeed.

File: tests/functional/shutdown/10-no-port-file/flow.cylc [hunk 7]
```diff
@@ -0,0 +1,13 @@
...
+        abort on timeout = True
+        timeout = PT1M
+
+[scheduling]
+    [[graph]]
+        R1 = t1
+
+[runtime]
+    [[t1]]
+        script = false
```
Comment (oliver-sanders) [inline tests/functional/shutdown/10-no-port-file/flow.cylc:13]: I don't think this should be here.

File: tests/functional/shutdown/11-bad-port-file/flow.cylc [hunk 8]
```diff
@@ -0,0 +1,13 @@
...
+        abort on timeout = True
+        timeout = PT1M
+
+[scheduling]
+    [[graph]]
+        R1 = t1
+
+[runtime]
+    [[t1]]
+        script = false
```
Comment (oliver-sanders) [inline tests/functional/shutdown/11-bad-port-file/flow.cylc:13]: I don't think this should be here.

File: tests/functional/validate/08-whitespace/flow.cylc [hunk 9]
```diff
@@ -0,0 +1,19 @@
...
+            bar
+            {% endif %}
+    [[   graph  ]]
+        T00, T12  ="""
+        a => b
+        foo => bar &  baz  """
+
+   [  runtime   ]
+    [[  foo,bar   ,  a,  b  ]]
+%include "inc.cylc"
```
Comment (oliver-sanders) [inline tests/functional/validate/08-whitespace/flow.cylc:19]: Can't believe I'm saying this but there is some trailing whitespace *missing* from this file!

File: tests/functional/deprecations/03-suiterc.t [hunk 10]
```diff
@@ -0,0 +1,46 @@
...
+
+TEST_NAME="${TEST_NAME_BASE}-validate"
+run_ok "${TEST_NAME}" cylc validate .
+
+TEST_NAME="${TEST_NAME_BASE}-register"
+run_ok "${TEST_NAME}" cylc register "${SUITE_NAME}"
+
+exists_ok "flow.cylc"
+
+purge_suite "${SUITE_NAME}"
```
Comment (oliver-sanders) [inline tests/functional/deprecations/03-suiterc.t:46]: I don't think this should be here?

File: tests/functional/shutdown/10-no-port-file/flow.cylc (hunk 7 above)
Comment (MetRonnie) [reply tests/functional/shutdown/10-no-port-file/flow.cylc:13]: Good spot, the deletion must have been accidentally overridden when merging master into the branch.

Took a bit of googling to figure out how to get `git log` to show me the files that were deleted on master:
```
git log --diff-filter=D --stat af1765ceeae5b921e181b59d988002d7c2528a74..master
```

File: tests/functional/deprecations/03-suiterc.t (hunk 10 above)
Comment (MetRonnie) [reply tests/functional/deprecations/03-suiterc.t:46]: I specifically created this test 😛

File: tests/functional/validate/08-whitespace/flow.cylc (hunk 9 above)
Comment (MetRonnie) [reply tests/functional/validate/08-whitespace/flow.cylc:19]: Is there some way of preventing whitespace tests passing if the whitespace gets stripped by mistake?

Comment (oliver-sanders) [reply tests/functional/validate/08-whitespace/flow.cylc]: Could potentially write out the `flow.cylc` file then run `sed -i 's/$/ /'` on it to add one whitespace char to the end of each line.

File: tests/functional/deprecations/03-suiterc.t (hunk 10 above)
Comment (oliver-sanders) [reply tests/functional/deprecations/03-suiterc.t:46]: Derp.

**Comment by oliver-sanders** [comment]
I think we are good to go!

**Pull request merged**